jerry-core/jerry-snapshot.c

```diff
@@ -630,3 +632,433 @@ jerry_exec_snapshot (const void *snapshot_p, /**< snapshot */
+  }
+
+  if (child_idx2 < num_of_nodes
+      && jerry_save_literals_compare (literals[max_idx], literals[child_idx2]))
+  {
+    max_idx = child_idx2;
+  }
+
+  return max_idx;
+} /* jerry_save_literals_heap_max */
+
+/**
+ * Helper function for a heapsort algorithm.
+ */
+static void
+jerry_save_literals_down_heap (ecma_string_t *literals[], /**< array of literals */
+                               lit_utf8_size_t num_of_nodes, /**< number of nodes */
+                               lit_utf8_size_t node_idx) /**< index of parent node */
+{
+  while (true)
+  {
+    lit_utf8_size_t max_idx = jerry_save_literals_heap_max (literals,
+                                                            num_of_nodes,
+                                                            node_idx,
+                                                            2 * node_idx + 1,
+                                                            2 * node_idx + 2);
+    if (max_idx == node_idx)
+    {
+      break;
+    }
+
+    ecma_string_t *swap  = literals[node_idx];
+    literals[node_idx] = literals[max_idx];
+    literals[max_idx] = swap;
+
+    node_idx = max_idx;
+  }
+} /* jerry_save_literals_down_heap */
+
+/**
+ * Helper function for a heapsort algorithm.
+ */
+static void
+jerry_save_literals_sort (ecma_string_t *literals[], /**< array of literals */
+                          lit_utf8_size_t num_of_literals) /**< number of literals */
+{
+  if (num_of_literals < 2)
+  {
+    return;
+  }
+
+  lit_utf8_size_t lit_idx = (num_of_literals - 2) / 2;
+
+  while (lit_idx <= (num_of_literals - 2) / 2)
+  {
+    jerry_save_literals_down_heap (literals, num_of_literals, lit_idx--);
+  }
+
+  for (lit_idx = 0; lit_idx < num_of_literals; lit_idx++)
+  {
+    const lit_utf8_size_t last_idx = num_of_literals - lit_idx - 1;
+
+    ecma_string_t *swap = literals[last_idx];
+    literals[last_idx] = literals[0];
+    literals[0] = swap;
+
+    jerry_save_literals_down_heap (literals, last_idx, 0);
+  }
+} /* jerry_save_literals_sort */
+
+/**
+ * Append characters to the specified buffer.
+ *
+ * @return the position of the buffer pointer after copy.
+ */
+static uint8_t *
+jerry_append_chars_to_buffer (uint8_t *buffer_p, /**< buffer */
+                              uint8_t *buffer_end_p, /**< the end of the buffer */
+                              const char *chars, /**< string */
+                              lit_utf8_size_t string_size) /**< string size */
+{
+  if (buffer_p > buffer_end_p)
+  {
+    return buffer_p;
+  }
+
+  const lit_utf8_size_t str_size = (string_size == 0) ? (lit_utf8_size_t) strlen (chars) : string_size;
+
+  if (buffer_p + str_size <= buffer_end_p)
+  {
+    strncpy ((char *) buffer_p, chars, str_size);
+
+    return buffer_p + str_size;
+  }
+
+  /* Move the pointer behind the buffer to prevent further writes. */
+  /* Distance from the buffer represents the unused bytes. */
+  return (buffer_end_p + (buffer_end_p - buffer_p));
```

Comment:
We shift the buffer pointer by the number of the unused bytes in the buffer instead of 1,
thus we can determine the number of the used bytes at the end of `jerry_parse_and_save_literals()`.